### ISSUE TITLE
Instrument LFO support (FMS and AMS)

### ISSUE THREAD

**Issue by marcb0t** [sikthehedgehog/Echo#13]
Is it possible to include support for FMS and AMS control to the FM instruments?

There are these things to take into account:

1. Master LFO controller for speed of frequency - 7 settings
2. FMS depth control - 7 settings
3. AMS depth control - 3 settings
4. Every operator has an on/off setting for allowing AMS for a given instrument.

I would appreciate this support in Echo. Thank you!

**Comment by sikthehedgehog** [comment]
One big problem with this is that the setting is global. That's going to be an issue regarding BGM and SFX (could always leave it enabled, the problem is the value of the LFO). Guess I'll just have to warn people to be careful with it.

For the record, the AMS bit can be stored in the FM instruments (and won't get filtered out), so you can "fake" LFO events using raw FM register writes meanwhile (see event $F8). It's not like BGM and SFX can be made to never clash on this one anyway =P

**Comment by marcb0t** [comment]
Thank you for your fast reply and consideration into this matter. Here are my thoughts:

1. One could simply make sound effects that don't use LFO.
2. Make a special setting that shuts off LFO for the duration of the sound effect.
3. Just be content with sound effects that use LFO to effect the music in some way.
4. Leave it up to the game and music programmers to worry about the "clashing" of SFX and BGM.

This is all apart of the fun, after all, isn't it?

But I will say, the LFO's are a very valid feature of the ym2612, that I find useful in some situations.

I'm willing to take on the challenge of meshing quality SFX with BGM together with such a unique feature!

I was going to use FMS instead of vibrato for the overworld music on a JRPG I'm working on. It really produced some sweet electric-guitar squeals.

And with over world music, there generally are no SFX, aside from a menu screen you can access... which in case I can use PSG for the SFX which will not be effected by LFO settings.

For battle scenes with SFX, I don't have a need for LFO settings.

So bottom line, the global nature of LFO settings is not a problem for me.

**Comment by sikthehedgehog** [comment]
> Make a special setting that shuts off LFO for the duration of the sound effect.

Then you prevent the case where LFO is used for music on channels not touched by sound effects.

> Leave it up to the game and music programmers to worry about the "clashing" of SFX and BGM.

Yeah, probably going to go this route, already doing this for some other things.

Just hoping nobody is stupid enough to ignore the clashes and then blame the programmer. Already had my fair share of dealing with people who put important instruments in the channels that can be clobbered by sound effects (because they aren't paying attention to that) and it drives me nuts (if you ever played Sky Sanctuary zone as Sonic you know what I mean).

> And with over world music, there generally are no SFX, aside from a menu screen you can access... which in case I can use PSG for the SFX which will not be effected by LFO settings.

LFO is triple enabled (global, per channel and per operator). Sound effects aren't required to use LFO. And in fact you can still use LFO on both music and sound effects as long as they use the same LFO value.

**Comment by marcb0t** [comment]
> Just hoping nobody is stupid enough to ignore the clashes and then blame the programmer.

Tell those stupid people to just "buzz off"! Don't lose any sleep over them.

Or if they give you problems, send them my way, and I'll set them straight! ;)  What I mean, is that I'll gladly show them what they're doing wrong in a constructive manner.

> Already had my fair share of dealing with people who put important instruments in the channels that can be clobbered by sound effects (because they aren't paying attention to that) and it drives me nuts (if you ever played Sky Sanctuary zone as Sonic you know what I mean).

Yes I "know what you mean". However, as an old school gamer, I have a different outlook/perspective. You see, these glitches and clashes, and poor priority placements of sound effects vs. lead instruments... this is all apart of the charm of retro chip music. Those imperfections and cutting out of things here and there in favor of sound effects. I actually like it when that happens, because it gave me the opportunity to focus more on the background track composition while the lead was silent for a few milliseconds. Some people are not so critical concerning these things. Sometimes, there is charm and value in the imperfections of a game.

Obviously, my goal is to make the most out of limitations, and I would certainly try to have SFX activate on a channel that will not cut out the lead instrument. But I don't find it to be such a disaster if there are a couple of clashes here or there.

> LFO is triple enabled (global, per channel and per operator). Sound effects aren't required to use LFO. And in fact you can still use LFO on both music and sound effects as long as they use the same LFO value.

Yes, I am familiar with how it all works. I'm very confident of my ability to use this feature responsibly and intuitively.

**Comment by sikthehedgehog** [comment]
> Yes I "know what you mean". However, as an old school gamer, I have a different outlook/perspective. You see, these glitches and clashes, and poor priority placements of sound effects vs. lead instruments... this is all apart of the charm of retro chip music. Those imperfections and cutting out of things here and there in favor of sound effects. I actually like it when that happens, because it gave me the opportunity to focus more on the background track composition while the lead was silent for a few milliseconds. Some people are not so critical concerning these things. Sometimes, there is charm and value in the imperfections of a game.

It's awful and pretty avoidable in our case, would prefer if people didn't do it >.> Letting a background bass cut out is OK, letting the instrument leading the melody is just plain awful.

Decide which channels the sound effects may take up and make sure to put whatever is safest to cut in them.

**Comment by marcb0t** [comment]
> It's awful and pretty avoidable in our case, would prefer if people didn't do it >.>

Oh yes, I agree. Certainly, it IS avoidable, especially in our case. WE don't have impossible deadlines, and have better and faster tools for testing these days. So no excuse for sacrificing a killer lead e-guitar. :)

I personally would choose a channel that is reserved for the lowest note of a chord, or a slow arpeggio line. Or PSG.

And this certainly gives me some additional things to consider. 

But yes, I think it's important to have full support of all the basic FM chip functions, which includes FMS and AMS. I've gotten some really sweet vibraphone instruments with AMS as well. They do come in handy from time to time.

And I really am glad that we are even having this conversation, and that you even listen to our requests and concerns. This musician appreciates all your hard work on the Echo driver. It is very much needed!

**Comment by sikthehedgehog** [comment]
I'm not sure deadlines are a factor really. Early on you arrange what channels are expected to be clobbered by sound effects. Then you would know ahead of time to not put anything important there (while also ensuring sound effects are only kept in those channels). It's all a matter of setting some guidelines early on and following them.

The only ones that get problematic are PCM and PSG noise since those are one of a kind. At least noise is usually only used for hit-hats (which aren't even the main percussion). PCM can be harder to work around... at least it's not used for the main melody. Unless you're Strobe (argh!).

**Comment by marcb0t** [comment]
> It's all a matter of setting some guidelines early on and **following them**.

Ah yes. True. But sometimes life is messy, and with strict deadlines, it's easy for people to be in such a rush that some things can "slip through the cracks", and mistakes can be made. But again, maybe I'm wrong.

Your suggestions are great, and I will certainly take them to heart.

I generally don't like to use PCM sound effects. I think PSG noise sound effects are generally kind of weak for most circumstances... I don't know... something for me to think about though.

> PCM can be harder to work around... at least it's not used for the main melody. Unless you're Strobe (argh!).

Oh really? How 'bout this:
https://soundcloud.com/marcb0t/marcb0t-house-demo

Or some of the improvisation leads and steel drums in this?:
https://soundcloud.com/marcb0t/ocean-melody-version-1

Ocean Melody is also the song where I made really convincing FM vibraphone instrument using AMS closer to the end of the song. If you want the module, I won't have a problem! ;)

**Comment by sikthehedgehog** [comment]
> I generally don't like to use PCM sound effects. I think PSG noise sound effects are generally kind of weak for most circumstances... I don't know... something for me to think about though.

Nah, PSG noise is pretty great for anything "whooshy". Like throwing a punch, or jumping (in a "realistic" setting), or a bird flapping its wings, etc. It *is* definitely hard - you need to be manipulating volume and pitch every tick (and likely use PSG3+PSG4 mode to control pitch accurately), but it's doable and once you learn to do that it's probably easier to do than FM noise.

Also this means you probably want to assign specifically PSG3 for square wave sound effects so you're already "wasting" it for a different reason.

> Oh really? How 'bout this:
> https://soundcloud.com/marcb0t/marcb0t-house-demo

That track is gone.

Speech is the one thing that's essentially unavoidable in PCM (in theory a normal FM channel can do vowels and the CSM mode in FM3 should allow for consonants, but yeaaaah good luck). Also by general rule I'm not really fond of FM drums :P I like the ones from Illumination Laser but I haven't figured out exactly what they are :/ (I doubt that dropping the second detune is a big deal, since in my experience detune doesn't affect sound that much)

Probably will be much less of an issue if #7 gets implemented, but that's... not gonna be easy at all since it'll eat cycles like crazy (ironically, leaving the PCM format as-is may be the best - in some test code I have it's actually doing 9-bit mixing since it's easier to just include the carry bit o_o).

**Comment by sikthehedgehog** [comment]
Anyway, back on topic:

* FMS and AMS are specified with register $B4+... which is the same register used for panning. The pan event ($F0-$F7) passes the register as-is, ergo that part is *already* covered (there's a reason why the official name is "set FM parameters").

* AM is the per-operator enable. This is in the same register as the first decay rate. Echo doesn't filter it out, so this part is already covered as well - just include it in the FM instrument itself. Note that when the docs say to leave as 0 any unused bits, it's referring to bits unused *by the YM2612*, not by Echo.

* LFO frequency is literally the only thing not "officially" supported by Echo. You can actually implement it as F8 22 xx (where xx = 00 to disable, or 08 to 0F to enable with a higher value meaning a higher frequency). Since LFO is global, there's no point to worry about not clashing with sound effects, so this should be already safe.

...tl;dr you literally an already do it _now_. It just doesn't have an official event and you're really losing at most one byte compared to a proper implementation (since I doubt you'll be messing with the LFO frequency all the time).

**Comment by marcb0t** [comment]
Oh wow! I did not realize all this. I will have to talk to my programmer so we can do some testing with this.

It looks like you covered mostly everything. However I do have a question concerning your comment here:

> AM is the per-operator enable. This is in the same register as the first decay rate. Echo doesn't filter it out, so this part is already covered as well - just include it in the FM instrument itself. Note that when the docs say to leave as 0 any unused bits, it's referring to bits unused by the YM2612, not by Echo.

So does this mean AMS setting 0-3 is also included in the instrument data as well? Or is there another register used for controlling the AMS depth?

Other than that everything else is fairly clear.

I will also pass this information on to the "dmf2esf" programmer to see if he can connect the Deflemask parameters to what you have mentioned here, into his conversion utility.

Thanks again!

**Comment by sikthehedgehog** [comment]
> So does this mean AMS setting 0-3 is also included in the instrument data as well? Or is there another register used for controlling the AMS depth?

AM is the per-operator enable, AMS is the sensivity. The former is part of the instrument, the latter would have to be set using the Set FM Params event (e.g. F0 xx for the first channel).

Specifically:
* Bits 7-6 for panning
* Bits 5-4 for AMS
* Bits 2-0 for FMS

(bit 3 is unused and should be left as 0)
